R/scale-discrete-.r

```diff
@@ -132,6 +134,14 @@ ScaleDiscretePosition <- ggproto("ScaleDiscretePosition", ScaleDiscrete,
     }
   },
 
+  get_breaks = function(self, limits = self$get_limits()) {
+    if (self$warn && length(self$get_limits()) == 0) {
```

Comment:
Having a function that is called between training and mapping seems most important — we could also use it to cache the palette for discrete scales.  Given that we need for two scales, it suggests it might be useful in other places.  Let's call it `train_complete()`.